### PR TITLE
[jit] Allow 'Any'  to appear as a type argument.

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26788 [jit] clean up exported source format
* #26767 [jit] Remove attempToRecoverType
* **#26572 [jit] Allow 'Any'  to appear as a type argument.**

Combined with isinstance specialization this allows a degree of polymorphic
functions to work without needing to use our weirder overload hacks.

We do not define any operators on Any, so the only thing you can do with it
is to put it in containers or type refine it using an isinstance check.
Any is restricted from appearing in non-argument position because we
cannot restore type tags if it ends up as a field in a class.

Differential Revision: [D17530643](https://our.internmc.facebook.com/intern/diff/D17530643)

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -6801,6 +6801,24 @@ def foo2(a : Optional[int], b : Optional[int]):
...
+        def any_refinement(a, b):
+            # type: (Any, Any) -> int
+            if isinstance(a, int) and isinstance(b, int):
+                return a + b
+            return 0
+
+        self.assertEqual(any_refinement(3, 4), 7)
+        self.assertEqual(any_refinement(3, "hi"), 0)
+
+    def test_any_in_class_fails(self):
```
Comment (suo) [inline test/test_jit.py]: Can we add tests that verify that `Any` is not allowed in Tuples, NamedTuples, or Interfaces as well?

File: aten/src/ATen/core/jit_type.h [hunk 2]
```diff
@@ -782,6 +782,19 @@ struct CAFFE2_API NamedType : public Type {
...
 };
 
+// Any should never appear in a named type like a class, namedtuple or
+// interface. If it does, then dynamic type information will be lost in the
+// Pickler, leading to hard-to-track-down bugs that will only occur
+// after saving or loading a model. This is because we rely on the
+// static types in named types to reconstruct type tags of loaded
+// values. Lifting this restriction requires solving the serialization
+// problem first.
+CAFFE2_API void checkNoAny(
```
Comment (suo) [inline aten/src/ATen/core/jit_type.h]: I am a little wary of having this enforcement as opt-in. For example, if someone goes in and adds attributes to interfaces and forgets to include this check, we've very quickly produced the bug this comment describes.

**Review by suo — APPROVED** [review]
ngnt

File: aten/src/ATen/core/jit_type.h (hunk 2 above)
Comment (zdevito) [reply aten/src/ATen/core/jit_type.h]: The specific example of interfaces will not actually cause this bug, but I agree in spirit. But how would we make it more likely to be called? There isn't a more common place to put it.

File: torch/csrc/jit/pybind_utils.h [hunk 3]
```diff
@@ -287,37 +285,24 @@ inline bool isTraceableType(TypePtr type) {
...
-    return TypedIValue(toIValue(input, type), type);
-  }
+  return toIValue(input, match.type());
+}
 
-  AT_ERROR(
+inline Stack toTraceableStack(const py::tuple& inputs) {
+  auto info = toTypeInferredIValue(inputs);
+  AT_CHECK(
+      isTraceableType(info.type()),
```
Comment (tangleintel) [inline torch/csrc/jit/pybind_utils.h:300]: Hi, @zdevito 
May I know why we only support Tensors and (possibly nested) Lists, Dicts, and Tuples of Tensors?
What if we passed a `None` type of python, and what will happen that makes us not support it?

**Pull request closed without merging**